Refactor: Rearchitect Usage Recording for Improved Reliability and Pe…

fraybyl1 opened the pull request:
This commit refactors the usage recording system, introducing a class-based
architecture for better organization and maintainability. 
**New Classes:**
`StatsCollector`: Fetches statistics from the XRay API
`DBManager`: Handles database interactions (MySQL/MariaDB/SQLITE), including deadlock retries.
`UsageRecorder`: Manages the recording process, including multithreaded data collection.

**Improved Error Handling:**

- More detailed error logging.
- Timeouts for XRay API calls.
- Database deadlock handling with retries.

**Optimized Database Operations:**

- Batched data processing.
- Efficient data update logic.

**Enhanced Concurrency:**

- Configurable thread pool (max_workers).
- Timeouts for thread execution.

**This refactoring improves code structure, robustness, and performance.**

We should probably put batch_size, max_workers in env. At least max_workers I think for sure. Batch_size should probably be calculated dynamically. I'm not sure. 

This is my last PR series for the near future. I hope these changes and [#1650](https://github.com/Gozargah/Marzban/pull/1650#issue-2848451408) [#1649 ](https://github.com/Gozargah/Marzban/pull/1649#issue-2848289559) [#67](https://github.com/Gozargah/Marzban-node/pull/67#issue-2850167282) will help in improving stability, responsiveness and performance with a large number of nodes and users. 

**And, of course, untested. Should be tested to see if it works. Thank you all.**

Avaritia-ZOV committed:
Refactor: Rearchitect Usage Recording for Improved Reliability and Performance

This commit refactors the usage recording system, introducing a class-based
architecture for better organization and maintainability. Key improvements:

New Classes:
StatsCollector: Fetches statistics from the XRay API.

DBManager: Handles database interactions (MySQL/MariaDB/SQLITE), including deadlock retries.

UsageRecorder: Manages the recording process, including multithreaded data collection.

Improved Error Handling:
More detailed error logging.
Timeouts for XRay API calls.
Database deadlock handling with retries.



Optimized Database Operations:
Batched data processing.
Efficient data update logic.

Enhanced Concurrency:
Configurable thread pool (max_workers).
Timeouts for thread execution.

This refactoring improves code structure, robustness, and performance.

ImMohammad20000 commented:
change target branch to dev

ghost commented:
Oops, didn't notice. I apologise

ImMohammad20000 commented on app/jobs/record_usages.py:
```diff
@@ -21,205 +22,366 @@
-        return
+    def process_user_stats(
+        self, user_stats: Dict, usage_coefficients: Dict
+    ) -> Generator[tuple[list[dict[str, int | Any]], dict[_KT, int]], None, None]:
```
```
Marzban/app/jobs/record_usages.py", line 243, in UsageRecorder
    ) -> Generator[tuple[list[dict[str, int | Any]], dict[_KT, int]], None, None]:
                                                          ^^^
NameError: name '_KT' is not defined
```

ImMohammad20000 reviewed (changes requested):
when you connected to vpn it can't record and save the user data 

```
ERROR:    Database error: bulk synchronize of persistent objects not supported when using bulk update with additional WHERE criteria right now.  add synchronize_session=None execution option to bypass synchronize of persistent objects.
ERROR:    Error in record_user_usages: bulk synchronize of persistent objects not supported when using bulk update with additional WHERE criteria right now.  add synchronize_session=None execution option to bypass synchronize of persistent objects.
Job "record_user_usages (trigger: interval[0:00:10], next run at: 2025-02-14 11:10:27 UTC)" raised an exception
Traceback (most recent call last):
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Marzban/app/jobs/record_usages.py", line 307, in record_user_usages
    DBManager.safe_execute(db, stmt, users_batch)
  File "/Marzban/app/jobs/record_usages.py", line 89, in safe_execute
    db.execute(stmt, params)
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2362, in execute
    return self._execute_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2247, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/bulk_persistence.py", line 1617, in orm_execute_statement
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: bulk synchronize of persistent objects not supported when using bulk update with additional WHERE criteria right now.  add synchronize_session=None execution option to bypass synchronize of persistent objects.
```

Avaritia-ZOV committed:
Small fix

My IDE decided to remove. 

from _typeshed import _KT

Avaritia-ZOV committed:
Fix: Database error: bulk synchronize

Avaritia-ZOV committed:
small fixes.

Avaritia-ZOV committed:
correct comparison of node_id with NULL.

ImMohammad20000 commented:
debugging this without you tested first is very hard try test it first

    ) -> Generator[tuple[list[dict[str, int | Any]], dict[_KT, int]], None, None]:
                                                          ^^^
NameError: name '_KT' is not defined. Did you mean: 'KT'?

Avaritia-ZOV committed:
Update record_usages.py

Avaritia-ZOV committed:
Chore: Keys are always int, no need for Generic.

I suppose we will never get a dictionary key different from INT, so generic is not required here.

ghost commented:
I think it's fixed now.

ImMohammad20000 commented on app/jobs/record_usages.py:
```diff
@@ -21,205 +22,372 @@
+
+                for attempt in range(3):
+                    try:
+                        db.execute(stmt, params)
```
use db.connection().execute(stmt, params)
also add it for line 91

fraybyl1 replied:
Is there a specific reason we're bypassing the db.execute() method and going directly to the raw connection?Maybe I'm wrong.

ImMohammad20000 replied:
In current impediment we are not use an id for some records but the db.execute method force us to use an id field so we need to bypass it

fraybyl1 replied:
I added `raw connection` for `inserts` to the table. It seems there should be no problems now.  What you think?

ImMohammad20000 replied:
i have a hard time to understand the difference
```python
if use_raw_connection:
                     conn.execute(stmt, params)
                else:
                    conn.execute(stmt, params)

```

ImMohammad20000 replied:
and this change still have errors
```
ERROR:    Database error: No primary key value supplied for column(s) users.id; per-row ORM Bulk UPDATE by Primary Key requires that records contain primary key values (Background on this error at: https://sqlalche.me/e/20/bupq)
ERROR:    Error in record_user_usages: No primary key value supplied for column(s) users.id; per-row ORM Bulk UPDATE by Primary Key requires that records contain primary key values (Background on this error at: https://sqlalche.me/e/20/bupq)
Job "record_user_usages (trigger: interval[0:00:10], next run at: 2025-02-14 12:38:52 UTC)" raised an exception
Traceback (most recent call last):
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Marzban/app/jobs/record_usages.py", line 323, in record_user_usages
    DBManager.safe_execute(db, stmt, users_batch)
  File "/Marzban/app/jobs/record_usages.py", line 100, in safe_execute
    conn.execute(stmt, params)
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2362, in execute
    return self._execute_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2247, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/bulk_persistence.py", line 1627, in orm_execute_statement
    result = _bulk_update(
             ^^^^^^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/bulk_persistence.py", line 357, in _bulk_update
    persistence._emit_update_statements(
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 781, in _emit_update_statements
    ) in groupby(
         ^^^^^^^^
  File "/.pyenv/versions/marzban/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 560, in _collect_update_commands
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: No primary key value supplied for column(s) users.id; per-row ORM Bulk UPDATE by Primary Key requires that records contain primary key values (Background on this error at: https://sqlalche.me/e/20/bupq)

```

fraybyl1 replied:
I really apologise, I don't know how I missed this. I apologise very much

ImMohammad20000 replied:
i think better to use Use raw connection true by default

fraybyl1 replied:
Hmm. Then we'll leave the raw connection for now. I apologise again for this indentation error. It's really awful.

M03ED commented:
these changes are so good and can help so much before we go to version 1.
currently were going to replace current [node](https://github.com/M03ED/gozargah-node/tree/dev) (check dev branch) with new version in v1, im trying to separate node code from marzban and put it in a library, [bridge-go](https://github.com/m03ed/gozargah_node_bridge) this is code in go and i will publish python code in next few days
can we use your help to check this ?

ghost commented:
> i will publish python code in next few days can we use your help to check this ?

Sure, I'll familiarise myself with the code. That's great news! It will solve so many problems!

Avaritia-ZOV committed:
fix: Use raw connection for inserts

Use raw connection (db.connection().execute()) for insert operations
in NodeUserUsage and NodeUsage tables to avoid issues with
missing auto-increment primary key.  Update operations and other
database interactions continue to use SQLAlchemy ORM (db.execute()).

Avaritia-ZOV committed:
Update record_usages.py

Avaritia-ZOV committed:
We'll use raw by default, but leave the option to switch easily.

ImMohammad20000 commented on app/jobs/record_usages.py:
```diff
@@ -81,7 +81,7 @@ def safe_execute(db: Session, stmt, params=None, use_raw_connection=False):
                 for attempt in range(3):
                     try:
                         if use_raw_connection:
```
remove if statement we dont need it

Avaritia-ZOV committed:
Getting rid of sleep deprivation delusions.

ImMohammad20000 reviewed (approved):
Its look good to me but we need @SaintShit opinion

M03ED commented:
here is the [bridge](https://pypi.org/project/gozargah-node-bridge/) library for python 
i write docs in next few days for it

M03ED commented:
i just realized something weird in code 
there is a method for `UsageRecorder `called `collect_stats` and you used that for getting both outbound stats and users stats
but in job after you called this method it just use one of them and throw other one.
when we call a xray api we need to reset traffic and when we do that if we don't record that usage its gone for ever
this can cause massive problem in recording users and outbounds usage

M03ED commented:
i don't think we even need `UsageRecorder` class
